File: app/src/main/java/com/left/gank/ui/cure/CurePresenter.java
package com.left.gank.ui.cure;

import android.business.domain.Gift;
import android.content.Context;
import android.jsoup.JsoupServer;
import android.rxbus.RxApiManager;
import android.text.TextUtils;
import android.ui.logcat.Logcat;

import com.left.gank.utils.CrashUtils;
import com.left.gank.utils.StringUtils;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.IllegalFormatException;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.atomic.AtomicBoolean;

import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

/**
 * Create by LingYan on 2016-10-26
 */

class CurePresenter extends CureContract.Presenter {
    private static final String BASE_URL = "http://www.mzitu.com/xinggan/";
    private final AtomicBoolean destroyFlag = new AtomicBoolean(false);

    CurePresenter(Context context, CureContract.View view) {
        super(context, view);
    }

    @Override
    void loadData(int page) {
        fetchData(page, getUrl(page));
    }

    @Override
    void loadImages(String url) {
        Disposable disposable = JsoupServer.rxConnect(url)
                .build()
                .doOnSubscribe(new Consumer<Disposable>() {
                    @Override
                    public void accept(Disposable disposable) throws Exception {
                        if (isViewLife()) {
                            view.showLoadingDialog();
                        }
                    }
                })
                .doFinally(() -> {
                    if (isViewLife()) {
                        view.hideLoadingDialog();
                    }
                })
                .subscribe(document -> {
                    getDetailMaxPage(document);
                    List<Gift> list = getImagesList(document);
                    if (isViewLife()) {
                        view.loadImagesSuccess(list);
                    }
                }, e -> {
                    if (isViewLife()) {
                        view.loadImagesFailure(errorTip);
                    }
                });
        RxApiManager.get().add(requestTag, disposable);
    }

    private void fetchData(int page, String url) {
        if (destroyFlag.get()) {
            return;
        }

        Disposable disposable = JsoupServer.rxConnect(url).build()
                .doOnSubscribe(dis -> {
                    if (view != null) {
                        view.showProgress();
                    }
                })
                .doFinally(() -> {
                    if (view != null) {
                        view.hideProgress();
                    }
                })
                .subscribe(document -> {
                    int maxPageNumber = getMaxPageNum(document);
                    List<Gift> gifts = getPageLists(document);
                    if (gifts != null) {
                        if (view != null) {
                            view.loadDataSuccess(page, maxPageNumber, gifts);
                        }
                    }
                }, Logcat::e);
        RxApiManager.get().add(requestTag, disposable);
    }

    /**
     * 解析每面，封面入口地址
     */
    private List<Gift> getPageLists(Document doc) {
        List<Gift> list = new ArrayList<>();
        if (doc == null) {
            return null;
        }
        Elements href = doc.select("#pins > li > a");
        Elements img = doc.select("#pins a img");
        Elements times = doc.select(".time");

        final int countSize = href.size();
        final int imgSize = img.size();
        final int size = countSize > imgSize ? imgSize : countSize;

        if (size > 0) {
            for (int i = 0; i < size; i++) {
                String imgUrl = img.get(i).attr("data-original");
                String title = img.get(i).attr("alt");
                String url = href.get(i).attr("href");
                String time = times.get(i).text();
                if (!TextUtils.isEmpty(imgUrl) && !TextUtils.isEmpty(url)) {
                    list.add(new Gift(imgUrl, url, time, title));
                }
            }
        }
        return list;
    }

    private int getMaxPageNum(Document doc) {
        int p = 0;
        if (doc != null) {
            Elements count = doc.select(".nav-links a[href]");
            int size = count.size();
            if (size > 0) {
                for (int i = size - 1; i >= 0; i--) {
                    String num = count.get(i).text();
                    if (StringUtils.isNumeric(num)) {
                        try {
                            return Integer.parseInt(num);
                        } catch (IllegalFormatException e) {
                            Logcat.e(e);
                            CrashUtils.crashReport(e);
                        }
                    }
                }
            }
        }
        return p;
    }

    private int getDetailMaxPage(Document doc) {
        if (doc != null) {
            Elements pages = doc.select(".pagenavi span");
            if (pages != null && !pages.isEmpty()) {

            }
        }
        return 0;
    }

    private List<Gift> getImagesList(Document doc) {
        List<Gift> imagesList = new ArrayList<>();
        if (doc != null) {
            Elements images = doc.select(".main-image a img");
            Elements lengths = doc.select(".pagenavi a");
            String spanLength = null;
            if (lengths != null && lengths.size() > 0) {
                Element element = lengths.get(lengths.size() - 2);
                if (element != null) {
                    spanLength = element.select("span").get(0).html();
                }
            }
            int length;
            try {
                length = spanLength == null ? -1 : Integer.valueOf(spanLength);
            } catch (IllegalFormatException e) {
                length = -1;
            }
            if (images != null && images.size() > 0) {
                String src = images.get(0).attr("src");
                if (!TextUtils.isEmpty(src)) {
                    int index = src.lastIndexOf(".");
                    if (index > 0) {
                        String start = src.substring(0, index);
                        start = start.substring(0, start.length() - 1);
                        String type = src.substring(index);
                        String sub = start.substring(0, start.length() - 1);
                        if (length != -1) {
                            for (int i = 1; i < length; i++) {
                                String page = i < 10 ? String.format(Locale.SIMPLIFIED_CHINESE, "0%d", i) : String.valueOf(i);
                                final String url = sub + page + type;
                                imagesList.add(new Gift(url));
                            }
                        }
                    }
                }
            }
        }
        return imagesList;
    }

    private String getUrl(int page) {
        return page == 1 ? BASE_URL : BASE_URL + "page/" + page;
    }

    @Override
    public void destroy() {
        if (destroyFlag.compareAndSet(false, true)) {
            RxApiManager.get().remove(requestTag);
        }
        super.destroy();
    }

    @Override
    protected void onDestroy() {

    }
}

File: app/src/main/java/com/left/gank/ui/cure/CurePresenter.kt
package com.left.gank.ui.cure

import android.business.domain.Gift
import android.content.Context
import android.jsoup.JsoupServer
import android.rxbus.RxApiManager
import android.text.TextUtils
import android.ui.logcat.Logcat
import com.left.gank.utils.CrashUtils
import com.left.gank.utils.StringUtils
import com.uber.autodispose.ObservableSubscribeProxy
import org.jsoup.nodes.Document
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Create by LingYan on 2016-10-26
 */

internal class CurePresenter(context: Context, view: CureContract.View) : CureContract.Presenter(context, view) {
    private val destroyFlag = AtomicBoolean(false)

    override fun loadData(page: Int) {
        fetchData(page, getUrl(page))
    }

    override fun loadImages(url: String) {
        val disposable = JsoupServer.rxConnect(url)
                .build()
                .doOnSubscribe {
                    if (isViewLife) {
                        view.showLoadingDialog()
                    }
                }
                .doFinally {
                    if (isViewLife) {
                        view.hideLoadingDialog()
                    }
                }
                .subscribe({ document ->
                    getDetailMaxPage(document)
                    val list = getImagesList(document)
                    if (isViewLife) {
                        view.loadImagesSuccess(list)
                    }
                }, { e ->
                    if (isViewLife) {
                        view.loadImagesFailure(errorTip)
                    }
                })
        RxApiManager.get().add(requestTag, disposable)
    }

    private fun fetchData(page: Int, url: String) {
        if (destroyFlag.get()) {
            return
        }

        val disposable = JsoupServer.rxConnect(url).build()
                .doOnSubscribe { dis ->
                    if (view != null) {
                        view.showProgress()
                    }
                }
                .doFinally {
                    if (view != null) {
                        view.hideProgress()
                    }
                }
                .`as`<ObservableSubscribeProxy<Document>>(bindLifecycle<Document>())
                .subscribe({ document ->
                    val maxPageNumber = getMaxPageNum(document)
                    val gifts = getPageLists(document)
                    if (gifts != null) {
                        if (view != null) {
                            view.loadDataSuccess(page, maxPageNumber, gifts)
                        }
                    }
                }, { Logcat.e(it) })
    }

    /**
     * 解析每面，封面入口地址
     */
    private fun getPageLists(doc: Document?): List<Gift>? {
        val list = ArrayList<Gift>()
        if (doc == null) {
            return null
        }
        val href = doc.select("#pins > li > a")
        val img = doc.select("#pins a img")
        val times = doc.select(".time")

        val countSize = href.size
        val imgSize = img.size
        val size = if (countSize > imgSize) imgSize else countSize

        if (size > 0) {
            for (i in 0 until size) {
                val imgUrl = img[i].attr("data-original")
                val title = img[i].attr("alt")
                val url = href[i].attr("href")
                val time = times[i].text()
                if (!TextUtils.isEmpty(imgUrl) && !TextUtils.isEmpty(url)) {
                    list.add(Gift(imgUrl, url, time, title))
                }
            }
        }
        return list
    }

    private fun getMaxPageNum(doc: Document?): Int {
        val p = 0
        if (doc != null) {
            val count = doc.select(".nav-links a[href]")
            val size = count.size
            if (size > 0) {
                for (i in size - 1 downTo 0) {
                    val num = count[i].text()
                    if (StringUtils.isNumeric(num)) {
                        try {
                            return Integer.parseInt(num)
                        } catch (e: IllegalFormatException) {
                            Logcat.e(e)
                            CrashUtils.crashReport(e)
                        }

                    }
                }
            }
        }
        return p
    }

    private fun getDetailMaxPage(doc: Document?): Int {
        if (doc != null) {
            val pages = doc.select(".pagenavi span")
            if (pages != null && !pages.isEmpty()) {

            }
        }
        return 0
    }

    private fun getImagesList(doc: Document?): List<Gift> {
        val imagesList = ArrayList<Gift>()
        if (doc != null) {
            val images = doc.select(".main-image a img")
            val lengths = doc.select(".pagenavi a")
            var spanLength: String? = null
            if (lengths != null && lengths.size > 0) {
                val element = lengths[lengths.size - 2]
                if (element != null) {
                    spanLength = element.select("span")[0].html()
                }
            }
            var length: Int
            try {
                length = if (spanLength == null) -1 else Integer.valueOf(spanLength)
            } catch (e: IllegalFormatException) {
                length = -1
            }

            if (images != null && images.size > 0) {
                val src = images[0].attr("src")
                if (!TextUtils.isEmpty(src)) {
                    val index = src.lastIndexOf(".")
                    if (index > 0) {
                        var start = src.substring(0, index)
                        start = start.substring(0, start.length - 1)
                        val type = src.substring(index)
                        val sub = start.substring(0, start.length - 1)
                        if (length != -1) {
                            for (i in 1 until length) {
                                val page = if (i < 10) String.format(Locale.SIMPLIFIED_CHINESE, "0%d", i) else i.toString()
                                val url = sub + page + type
                                imagesList.add(Gift(url))
                            }
                        }
                    }
                }
            }
        }
        return imagesList
    }

    private fun getUrl(page: Int): String {
        return if (page == 1) BASE_URL else BASE_URL + "page/" + page
    }

    override fun destroy() {
        if (destroyFlag.compareAndSet(false, true)) {
            RxApiManager.get().remove(requestTag)
        }
        super.destroy()
    }

    override fun onDestroy() {

    }

    companion object {
        private val BASE_URL = "http://www.mzitu.com/xinggan/"
    }
}
